sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/sortBasedIterators.scala

```diff
@@ -409,10 +409,19 @@ class FinalSortAggregationIterator(
 
   override protected def initialInputBufferOffset: Int = groupingExpressions.length
 
+  private[this] var copyBuffer = new Array[Byte](64)
+  private def cheapRowCopy(row: InternalRow): UnsafeRow = {
+    val unsafeRow = row.asInstanceOf[UnsafeRow]
+    if (unsafeRow.getSizeInBytes() > copyBuffer.length) {
+      copyBuffer = new Array[Byte](unsafeRow.getSizeInBytes)
+    }
+    unsafeRow.copyTo(copyBuffer)
+  }
+
   override def initialize(): Unit = {
     if (inputIter.hasNext) {
       initializeBuffer()
-      val currentRow = inputIter.next().copy()
+      val currentRow = cheapRowCopy(inputIter.next())
```

Comment:
this is for demonstration, I'm not sure how to use this feature in aggregate sort.